File: ios/Native/Bridge/ChannelBridge.swift
//
//  ChannelBridge.swift
//  Runner
//
//  Created by Rafael Takahashi on 27/03/23.
//

import Foundation

/// Simple handler that uses blocking execution. Parameters are received from Flutter
/// and the return (or the thrown value) is returned to Flutter.
typealias SimpleBridgeHandler = (Any?) throws -> Any

/// Parameter 1: Parameters received from Flutter.
/// Parameter 2: Function to be called when fulfilling the callback.
/// Parameter 3: Function to be called when finishing with an error.
typealias NonblockingBridgeHandler = (Any?, @escaping (Any?) -> Void, @escaping (Any?) -> Void) -> Void

enum BridgeHandler {
    case simple(SimpleBridgeHandler)
    case nonblocking(NonblockingBridgeHandler)
}

/// iOS side of the method channel bridge.
///
/// Use this bridge to communicate with Flutter code instead of directly opening method channels.
/// To use it, open a port with a name.
///
/// After having obtained a port, you can use it just like you would use the method channel, with a few differences:
/// - Handlers are registered one by one, by method name. You shouldn't register a method name that has been registered before on
/// the same port. If you do, the previous handler will be discarded and that may lead to bugs.
/// - Port names work like prefixes, but you don't have to worry about constructing the names of methods.
/// - Calls and responses are exposed through simple callbacks that are easier to use than the method channel.
///
/// ```Swift
/// let weatherPort = MethodChannelBridge.openPort("WeatherRepository")
///
/// // Send a call to the Flutter side and get its response:
/// weatherPort.call(
///   "fetchWeatherForecast",
///   withArguments: args,
///   onError: { exception in logError(exception) }
///   onSuccess: { params in ... }
///   )
///
/// // or register to receive calls from the Flutter side:
/// weatherPort.registerHandler("fetchWeatherForecast") { (params) -> Any in ... }
/// // if you need a non-blocking handler, use the following overload:
/// weatherPort.registerHandler("fetchWeatherForecast") { (params, fulfill, reject) -> Void in ... }
/// ```
class MethodChannelBridge {
    /// This exact string must also be used in the Flutter side of the bridge.
    static let methodChannelName = "br.com.rtakahashi.playground.flutter_reference"
    
    /// A static method channel that's shared by all ports in the bridge.
    static var sharedChannel: FlutterMethodChannel? = nil
    
    /// Handlers for the iOS side of the method channel bridge.
    static var bridgeHandlers: Dictionary<String,BridgeHandler> = [:]
    
    // The functions in this class could've been top-level functions and they would work the same;
    // I decided to put them in a class to make their usage more explicit about what they are.
    // That is, you call them with MethodChannelBridge.initialize() rather than just calling
    // the function.
    
    private init() { /** Do not instantiate this class. */}
    
    /// Open a port to the Flutter side.
    ///
    /// Anything from navigation to repositories can use a port.
    ///
    /// "Open" might not be accurate because a port simply works as a prefix when sending and receiving messages through
    /// the message channel. Many classes may "open" the same method channel port and they'll all be able to use it, but you
    /// should avoid that because only one handler can be registered for each name in the same port.
    static func openPort(_ portName: String) -> MethodChannelBridgePort {
        return MethodChannelBridgePortImpl(portName: portName)
    }
    
    /// Initialize this side of the bridge. This function MUST be called at least once, informing the Flutter binary messenger.
    /// The bridge can be used to make calls to the Flutter side only after initialization.
    ///
    /// It is possible to register handlers before initialization, but method calls from the Flutter side will only reach this side of
    /// the bridge after initialization.
    static func initialize(withBinaryMessenger messenger: FlutterBinaryMessenger) {
        if (sharedChannel !== nil) {
            // Method channel already initialized. Skipping.
            return
        }
        
        let newChannel = FlutterMethodChannel(name: methodChannelName, binaryMessenger: messenger, codec: FlutterJSONMethodCodec())
        
        sharedChannel = newChannel
        
        // Set the one and only handler on the shared method channel. This handler checks the
        // method name (which also contains the port name) and routes it to the correct
        // bridge handler.
        newChannel.setMethodCallHandler() {
            (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            // All calls that arrive here are coming from the other side of the bridge.
            // The method names there are constructed using the same rules as this side.
            // Therefore, we expect the method name to include the port name.
            // Our handlers in the bridge are kept in a dictionary, where the keys are strings
            // that also follow that rule.
            let methodName = call.method
            guard let handler = bridgeHandlers[methodName] else {
                result(FlutterMethodNotImplemented)
                return
            }
            
            switch handler {
            case .simple(let simpleHandler):
                do {
                    // This value will be whatever is returned by the handler that was registered
                    // in the bridge. It may return nothing, in which case the Flutter side will
                    // receive a null.
                    let val = try simpleHandler(call.arguments)
                    // In case the handler returns void, we'll get an empty tuple.
                    // Check if that's the case, and if so, replace it with nil. Otherwise, the
                    // empty tuple causes a crash when the method handler tries to serialize it.
                    result(val is () ? nil : val)
                } catch let error {
                    result(FlutterError(code: "bridge-handler-execution-error", message: "iOS handler for method \(methodName) failed.", details: error.localizedDescription))
                }
            case .nonblocking(let nonblockingHandler):
                nonblockingHandler(call.arguments, {val in
                    // When the handler fulfills:
                    result(val is () ? nil : val)
                    // This callback is called manually, so we don't expect () as a possible
                    // return, but we check for it anyway just for safety.
                }, {err in
                    // When the handler rejects:
                    result(FlutterError(code: "bridge-handler-execution-error", message: "iOS handler for method \(methodName) failed.", details: "\(String(describing: err))"))
                })
                // "result" is not supposed to be called multiple times. This bridge doesn't have
                // any such enforcement of its own.
            }
            
            return
        }
    }
    
    /// Implementation of the method bridge port. See the protocol to see how to use a port.
    fileprivate class MethodChannelBridgePortImpl: MethodChannelBridgePort {
        
        
        private let portName: String
        
        fileprivate init(portName: String) {
            self.portName = portName
        }
        
        func call(_ methodName: String) {
            return call(methodName, withArguments: nil)
        }
        
        func call(_ methodName: String, withArguments args: Any?) {
            return call(methodName, withArguments: args) { _ in }
        }
        
        func call(_ methodName: String, withArguments args: Any?, onError errorCallback: @escaping (MethodChannelBridgeException) -> Void) {
            return call(methodName, withArguments: args, onError: errorCallback) { _ in }
        }
        
        func call(_ methodName: String, withArguments args: Any?, onError errorCallback: @escaping (MethodChannelBridgeException) -> Void, onSuccess successCallback: @escaping (Any?) -> Void) {
            // I considered making this an async function, but requiring iOS 13 just for that
            // doesn't seem worth it.
            
            // If the channel is still nil, that can only mean the bridge hasn't been initialized yet.
            guard let channel = MethodChannelBridge.sharedChannel else {
                errorCallback(MethodChannelBridgeException(MethodChannelBridgeExceptionCause.bridgeNotInitialized))
                return
            }
            
            // Invoke the method using the method channel.
            // We'll call either the error callback or the success callback depending on the
            // value.
            channel.invokeMethod("\(portName).\(methodName)", arguments: args) {
                (value: Any?) -> Void in
                if (value is FlutterError) {
                    errorCallback(MethodChannelBridgeException(MethodChannelBridgeExceptionCause.methodExecutionFailed))
                }
                else if (FlutterMethodNotImplemented.isEqual(value)) {
                    errorCallback(MethodChannelBridgeException(MethodChannelBridgeExceptionCause.receiverNotFound))
                } else {
                    successCallback(value)
                }
            }
        }
        
        func registerHandler(_ handlerName: String, _ handler: @escaping SimpleBridgeHandler) {
            let fullHandlerName = "\(portName).\(handlerName)"
            if (bridgeHandlers[fullHandlerName] != nil) {
                NSLog("Method \(handlerName) has already been registered for port \(portName) in the bridge, and will be discarded!")
            }
            
            bridgeHandlers[fullHandlerName] = BridgeHandler.simple(handler)
        }
        
        func registerHandler(_ handlerName: String, _ handler: @escaping NonblockingBridgeHandler) {
            let fullHandlerName = "\(portName).\(handlerName)"
            if (bridgeHandlers[fullHandlerName] != nil) {
                NSLog("Method \(handlerName) has already been registered for port \(portName) in the bridge, and will be discarded!")
            }
            
            bridgeHandlers[fullHandlerName] = BridgeHandler.nonblocking(handler)
        }
        
        func unregisterHandler(_ handlerName: String) {
            bridgeHandlers.removeValue(forKey: handlerName)
        }
        
        func unregisterAllHandlers() {
            let filtered = bridgeHandlers.filter() {
                $0.key.hasPrefix("\(portName).")
            }
            bridgeHandlers.removeAll()
            filtered.forEach { (key: String, value: BridgeHandler) in
                bridgeHandlers[key] = value
            }
        }
    }
}


/// Represents a port in our method channel bridge that can communicate with the Flutter side.
///
/// A port can be used to send data to the Flutter side, and also register methods that the Flutter side can call.
///
/// You can think of each port as only being able to communicate with the corresponding port (with the same name)
/// on the other side.
protocol MethodChannelBridgePort: AnyObject {
    
    /// Make a method call without arguments and without callbacks.
    /// Note that even though you're not receiving the result of this call, it's a blocking operation that will only complete
    /// when the corresponding method on the other side of the bridge finishes handling the call.
    ///
    /// Do consider also specifying an error callback whenever possible.
    func call(_ methodName: String) -> Void
    /// Make a method call with arguments (which may be null), but without callbacks.
    /// Note that even though you're not receiving the result of this call, it's a blocking operation that will only complete
    /// when the corresponding method on the other side of the bridge finishes handling the call.
    ///
    /// Do consider also specifying an error callback whenever possible.
    func call(_ methodName: String, withArguments args: Any?) -> Void
    /// Make a method call with arguments (which may be null) and an error handler, which will be called once if the call
    /// fails. It is highly recommended to always specify an error callback.
    func call(_ methodName: String, withArguments args: Any?, onError errorCallback: @escaping (MethodChannelBridgeException) -> Void)
    /// Make a method call with arguments (which may be null) and handlers for error and success. Either one of the
    /// handlers will be called exactly once.
    ///
    /// There must be a port on the Flutter side with the same name as this port with a registered handler with the method
    /// name specified in this method.
    ///
    /// If a corresponding port with a corresponding method handler doesn't exist in the Flutter side, or if either side of the bridge
    /// hasn't finished initializing, the error callback will be called receiving an exception as parameter. Otherwise, the success
    /// callback will be called receiving what the Flutter method returned (may be null).
    func call(_ methodName: String, withArguments args: Any?, onError errorCallback: @escaping (MethodChannelBridgeException) -> Void, onSuccess successCallback: @escaping (Any?) -> Void)
    
    /// Register a handler for a name. If a different handler was already registered for the same name, it'll be discarded
    /// in favor of this one.
    ///
    /// Note that this does not directly register a handler in the method channel. This handler is stored on the bridge, and
    /// when some message comes through the method channel, the bridge routes it to the correct handler.
    ///
    /// This overload registers a BLOCKING handler that must not be used if you intend to make api calls or any other asynchronous operation. Use the overload for a non-blocking handler instead for those cases.
    ///
    /// Handlers are allowed to throw, but avoid that because anything that's thrown gets turned into an error that's sent to
    /// the Flutter side of the bridge.
    ///
    /// Anything that's returned by the handler reaches the other side of the bridge, but the return value must be void or a
    /// serializable value.
    ///
    /// Usage:
    /// ```swift
    /// bridgePort.registerHandler("name") {
    ///   (params) in
    ///   // (Anything done here blocks the thread)
    ///   return value
    /// }
    /// ```
    func registerHandler(_ handlerName: String, _ handler: @escaping SimpleBridgeHandler)
    
    /// Register a handler for a name. If a different handler was already registered for the same name, it'll be discarded
    /// in favor of this one.
    ///
    /// Note that this does not directly register a handler in the method channel. This handler is stored on the bridge, and
    /// when some message comes through the method channel, the bridge routes it to the correct handler.
    ///
    /// This is a non-blocking overload that may be more difficult to use but is necessarily for long operations in order to not
    /// block the thread.
    /// The first parameter of the handler is for the parameters received from Flutter. The second and third parameters are
    /// the fulfill and reject functions, respectively; exactly one of the two must be called eventually, and only once, with the
    /// value to be returned to the Flutter side. It's important to fulfill with nil even if you don't return anything from the handler,
    /// so that the Flutter side doesn't wait for a response indefinitely.
    ///
    /// Usage:
    /// ```swift
    /// bridgePort.registerHandler("name") {
    ///   (params, fulfill, reject) in
    ///   // (Do asynchronous work)
    ///   fulfill(value)
    ///   // Call either fulfill or reject, and only once.
    /// }
    /// ```
    func registerHandler(_ handlerName: String, _ handler: @escaping NonblockingBridgeHandler)
    
    /// Unregister a handler by name.
    func unregisterHandler(_ handlerName: String)
    
    /// Unregister all handlers for this port.
    func unregisterAllHandlers()
}

/// Custom exception that represents some problem when calling methods on the Flutter side.
class MethodChannelBridgeException: NSObject, LocalizedError {
    public let cause: MethodChannelBridgeExceptionCause
    
    fileprivate init(_ cause: MethodChannelBridgeExceptionCause) {
        self.cause = cause
    }
    
    public override var description: String { return "Bridge error with cause \(cause)" }
    
    var errorDescription: String? {
        get {
            return self.description
        }
    }
}

enum MethodChannelBridgeExceptionCause {
    case receiverNotFound
    case bridgeNotInitialized
    case methodExecutionFailed
}

File: android/app/src/main/java/br/com/rtakahashi/playground/flutter_reference/core/bridge/ChannelBridge.kt
package br.com.rtakahashi.playground.flutter_reference.core.bridge

import android.util.Log
import br.com.rtakahashi.playground.flutter_reference.core.domain.error.InteropException
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.JSONMethodCodec
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

typealias SimpleBridgeHandlerFunction = (parameters: Any?) -> Any?
typealias NonblockingBridgeHandlerFunction = (parameters: Any?, fulfill: (Any?) -> Unit, reject: (Any?) -> Unit) -> Unit
typealias SuspendBridgeHandlerFunction = suspend (parameters: Any?) -> Any?

/**
 * We allow for different types of handlers to be registered in the bridge.
 * Each one needs to be called in a different way.
 */
private sealed interface BridgeHandler

/**
 * The simplest type of handler to write and call, but also blocks the thread
 * when it runs.
 */
private data class SimpleBridgeHandler(
    val f: SimpleBridgeHandlerFunction
) : BridgeHandler

/**
 * A non-blocking callback-based handler that may be harder to use but doesn't
 * block the thread when it runs.
 */
private data class NonblockingBridgeHandler(
    val f: NonblockingBridgeHandlerFunction
) : BridgeHandler

/**
 * A suspend handler that's the recommended way of registering handlers.
 * It's easy to use and doesn't require manually writing any callbacks,
 * but the other options are provided in case it causes concurrency problems.
 */
private data class SuspendBridgeHandler(
    val f: SuspendBridgeHandlerFunction
) : BridgeHandler

/**
 * Android side of the method channel bridge.
 *
 * Use this bridge to communicate with Flutter code instead of directly opening method channels.
 * To use it, open a port with a name.
 *
 * After having obtained a port, you can use it just like you would use the method channel, with a
 * few differences:
 * - Handlers are registered one by one, by method name. You shouldn't register a method name that
 * has been registered before on the same port. If you do, the previous handler will be discarded
 * and that may lead to bugs.
 * - Port names work like prefixes, but you don't have to worry about constructing the names of
 * methods.
 * - Calls and responses are exposed through simple callbacks that are easier to use than the
 * method channel.
 *
 * ```kotlin
 * val weatherPort = MethodChannelBridge.openPort("WeatherRepository");
 *
 * // Send a call to the Flutter side and get its response:
 * val forecast = weatherPort.call("fetchWeatherForecast", args, { logError(it) } ) { handleResponse(it) };
 * // or register to receive calls from the Flutter side:
 * weatherPort.registerHandlerSuspend("fetchWeatherForecast") suspend { params -> handleCall(params) };
 *
 * // if you have concurrency problems, you can use the other functions as well:
 * weatherPort.registerHandlerBlocking("fetchWeatherForecast") { params -> handleCall(params) };
 * weatherPort.registerHandlerPromise("fetchWeatherForecast") { params, fulfill, reject ->
 *  ...
 *  fulfill(value)
 * };
 * ```
 */
object MethodChannelBridge {
    // The functions in this object could've been top-level functions and they would work the same;
    // I decided to put them in an object to make their usage more explicit about that they are.
    // That is, you call them with MethodChannelBridge.initialize() rather than just calling
    // the function.

    /**
     * A static method channel that's shared by all ports in the bridge.
     */
    private var sharedChannel: MethodChannel? = null

    /**
     * Handlers for the Android side of the method channel bridge.
     */
    private var bridgeHandlers: MutableMap<String, BridgeHandler> = mutableMapOf()

    /**
     * This exact string must also be used in the Flutter side of the bridge.
     */
    private const val methodChannelName = "br.com.rtakahashi.playground.flutter_reference"

    /**
     * Open a port to the Flutter side.
     *
     * Anything from navigation to repositories can use a port.
     *
     * "Open" might not be accurate because a port simply works as a prefix when sending and
     * receiving messages through the method channel. Many classes may "open" the same method
     * channel port and they'll all be able to use it, but you should avoid that because only one
     * handler can be registered for each name in the same port.
     */
    @JvmStatic
    fun openPort(portName: String): AndroidMethodChannelBridgePort {
        // It is better to "open" method channels using a method rather than letting other
        // classes directly instantiate ports because we may want to add some more control
        // here in the future.
        if (portName.isEmpty() || portName.contains('.')) {
            throw IllegalArgumentException(
                "$portName is not a valid port name;" +
                        " it must not be empty and must not contain any dots ('.') in it."
            )
        }
        return AndroidMethodChannelBridgePortImpl(portName)
    }


    /**
     * Initialize this side of the bridge. This function MUST be called at least once, informing
     * the Flutter engine. The bridge can be used to make calls to the Flutter side only after
     * initialization.
     *
     * It is possible to register handlers before initialization, but method calls from the
     * Flutter side will only reach this side of the bridge after initialization.
     */
    @JvmStatic
    fun initialize(engine: FlutterEngine) {
        if (sharedChannel !== null) {
            // Method channel already initialized. Skipping.
            return
        }

        val newChannel = MethodChannel(
            engine.dartExecutor.binaryMessenger,
            methodChannelName, JSONMethodCodec.INSTANCE
        )

        newChannel.setMethodCallHandler { call, result ->
            // All calls that arrive here are coming from the other side of the bridge.
            // The method names there are constructed using the same rules as this side.
            // Therefore, we expect the method name to include the port name.
            // Our handlers in the bridge are kept in a map, where the keys are strings that also
            // follow that rule.
            val methodName = call.method

            when (val handler = bridgeHandlers[methodName]) {
                null -> result.notImplemented()
                is SimpleBridgeHandler -> {
                    // A simple bridge handler is one that was registered as a simple function
                    // of Any? to Any?, and it blocks the thread while it runs.
                    try {
                        // Execute the handler and send the result back through the method channel.
                        result.success(handler.f.invoke(call.arguments))
                    } catch (e: Exception) {
                        errorWithException(result, e)
                    }
                }

                is NonblockingBridgeHandler -> {
                    // A non-blocking handler is one that receives the parameters and also a fulfill
                    // and a reject callback, and it's expected to call one of them.
                    handler.f.invoke(call.arguments, {
                        result.success(it)
                    }, {
                        errorWithException(result, it)
                    })
                }

                is SuspendBridgeHandler -> {
                    // A suspend handler needs to execute in the global scope.
                    GlobalScope.launch {
                        try {
                            val res = handler.f.invoke(call.arguments)
                            result.success(res)
                        } catch (e: Exception) {
                            errorWithException(result, e)
                        }
                    }
                }
            }

        }

        sharedChannel = newChannel
    }

    // Call the method channel's result with the information contained in the exception.
    private fun errorWithException(methodChannelResult: MethodChannel.Result, e: Any?) {
        if (e is InteropException) {
            // In this case, we have additional information to send.
            methodChannelResult.error(
                e.errorCode,
                e.message,
                e.extra
            );
        } else if (e is Exception) {
            methodChannelResult.error(
                "bridge-handler-execution-error",
                "Android handler has failed with exception: $e",
                e.toString(),
            );
        } else {
            methodChannelResult.error(
                "bridge-handler-execution-error",
                "Android handler has failed with unknown error: $e",
                e?.toString(),
            );
        }
    }


    // (This is a private class that implements AndroidMethodChannelBridgePort in order to prevent
    // it from being instantiated elsewhere. See the interface for how to use a port.)
    private class AndroidMethodChannelBridgePortImpl(val name: String) :
        AndroidMethodChannelBridgePort {

        override fun call(
            methodName: String,
            arguments: Any?,
            errorCallback: ((MethodChannelBridgeException) -> Any)?,
            callback: ((Any?) -> Any)?
        ) {
            // This variable is just to enable a smart cast.
            val channel = sharedChannel

            if (channel != null) {
                channel.invokeMethod("$name.$methodName", arguments, object : MethodChannel.Result {
                    override fun success(result: Any?) {
                        callback?.invoke(result)
                    }

                    override fun notImplemented() {
                        errorCallback?.invoke(MethodChannelBridgeException.receiverNotFound())
                    }

                    override fun error(
                        errorCode: String,
                        errorMessage: String?,
                        errorDetails: Any?
                    ) {
                        errorCallback?.invoke(MethodChannelBridgeException.methodExecutionFailed())
                    }
                })
            } else {
                errorCallback?.invoke(MethodChannelBridgeException.bridgeNotInitialized())
            }
        }

        override suspend fun call(methodName: String, arguments: Any?): Any? {
            // When calling the method channel from Kotlin, it's important to do so from the
            // main thread, otherwise the call simply will never complete (the callbacks will never
            // be called).
            // Simply making this a coroutine with `suspendCoroutine` would not ensure that the
            // call happens in the method channel, so the call to `withContext` around it is needed.
            return withContext(Dispatchers.Main) {
                suspendCoroutine { continuation ->
                    call(methodName, arguments, { error ->
                        continuation.resumeWithException(error)
                    }, { result ->
                        continuation.resume(result)
                    })
                }
            }
        }

        override fun registerHandlerBlocking(
            callbackName: String,
            handler: (SimpleBridgeHandlerFunction)
        ) {
            _registerHandler(callbackName, SimpleBridgeHandler(handler))
        }

        override fun registerHandlerPromise(
            callbackName: String,
            handler: NonblockingBridgeHandlerFunction
        ) {
            _registerHandler(callbackName, NonblockingBridgeHandler(handler))
        }

        override fun registerHandlerSuspend(
            callbackName: String,
            handler: SuspendBridgeHandlerFunction
        ) {
            _registerHandler(callbackName, SuspendBridgeHandler(handler))
        }

        private fun _registerHandler(callbackName: String, handler: BridgeHandler) {
            if (bridgeHandlers["$name.$callbackName"] !== null) {
                Log.d(
                    "bridge",
                    "Method $callbackName for port $name has already been registered in the bridge!"
                )
                Log.d("bridge", "Previous handler for $name.$callbackName will be discarded.")
            }

            bridgeHandlers["$name.$callbackName"] = handler
        }

        override fun unregisterHandler(callbackName: String) {
            bridgeHandlers.remove(callbackName)
        }

        override fun unregisterAllHandlers() {
            bridgeHandlers =
                bridgeHandlers.filter { entry -> entry.key.startsWith("$name.") }.toMutableMap()
        }

    }
}


/**
 * Represents a port in our method channel bridge that can communicate with the Flutter side.
 *
 * A port can be used to send data to the Flutter side, and also register methods that the Flutter
 * side can call. Both a callback-based and a suspend version of the `call` method are provided,
 * so that callers can choose to use straightforward try/catches with suspend functions and use
 * callbacks from non-suspend functions if necessary.
 *
 * You can think of each port as only being able to communicate with the corresponding port (with
 * the same name) on the other side.
 */
interface AndroidMethodChannelBridgePort {

    /**
     * Send something through the shared channel.
     *
     * There must be a port on the Flutter side with the same name as this port with a registered
     * handler with the method name specified as [methodName].
     *
     * If a corresponding port with a corresponding method handler doesn't exist in the Flutter
     * side, or if either side of the method channel bridge hasn't finished initializing, the error
     * callback will be called receiving an exception as parameter. Otherwise, the success callback
     * will be called receiving what the Flutter method returned (may be null).
     *
     * The success callback is declared last for two reasons:
     * 1. Both callbacks are optional, but it's highly recommended to always handle the error case.
     * 2. Having the success callback declared last allows writing it outside the parentheses when
     * calling this method.
     *
     * The return type of the callbacks does not matter. They're given as "Any" to allow for not
     * returning anything.
     */
    fun call(
        methodName: String,
        arguments: Any? = null,
        errorCallback: ((MethodChannelBridgeException) -> Any)? = null,
        callback: ((Any?) -> Any)? = null
    )

    /**
     * Send something through the shared channel.
     *
     * This works the same as the callback-based overload, except that it's a suspend function.
     */
    suspend fun call(methodName: String, arguments: Any? = null): Any?

    /**
     * Register a handler for a name. If a different handler was already registered for the same
     * name, it'll be discarded in favor of this one.
     *
     * Note that this does not directly register a handler in the method channel. This handler is
     * stored on the bridge, and when some message comes through the method channel, the bridge
     * routes it to the correct handler.
     *
     * It is always recommended to call `registerHandlerSuspend()`, which accepts a suspend function.
     * This function registers a BLOCKING handler that must not be used if you intend to make api
     * calls or other asynchronous operation.
     *
     * Handlers are allowed to throw, but avoid that because anything that's thrown gets turned
     * into an error that's send to the Flutter side of the bridge.
     *
     * Usage:
     * ```kotlin
     * bridgePort.registerHandlerBlocking("name") { params ->
     *   ...
     *   return someValue
     *   // Or throw when there's an error.
     * }
     * ```
     */
    fun registerHandlerBlocking(callbackName: String, handler: SimpleBridgeHandlerFunction)

    /**
     * Register a handler for a name. If a different handler was already registered for the same
     * name, it'll be discarded in favor of this one.
     *
     * Note that this does not directly register a handler in the method channel. This handler is
     * stored on the bridge, and when some message comes through the method channel, the bridge
     * routes it to the correct handler.
     *
     * This is a non-blocking method that may be more difficult to use, but unlike the
     * `registerHandlerBlocking()` function, it doesn't block the thread. However, do consider
     * using `registerHandlerSuspend()` instead of this.
     *
     * The first parameter of the handler is for the parameters received from Flutter. The second
     * and third parameters are the fulfill and reject functions, respectively; exactly one of the
     * two must be called eventually, and only once, with the value to be returned to the Flutter
     * side. It's important to fulfill with null even if you don't return anything from the handler,
     * so that the Flutter side doesn't wait for a response indefinitely.
     *
     * Usage:
     * ```kotlin
     * bridgePort.registerHandlerPromise("name") { params, fulfill, reject ->
     *   ...
     *   fulfill(someValue)
     *   // Call either fulfill or reject, only once.
     * }
     * ```
     */
    fun registerHandlerPromise(
        callbackName: String,
        handler: NonblockingBridgeHandlerFunction
    )

    /**
     * Register a handler for a name. If a different handler was already registered for the same
     * name, it'll be discarded in favor of this one.
     *
     * Note that this does not directly register a handler in the method channel. This handler is
     * stored on the bridge, and when some message comes through the method channel, the bridge
     * routes it to the correct handler.
     *
     * There are other versions of this function that you can use if you encounter concurrency
     * problems. However, this is the recommended way of registering handlers in the bridge.
     *
     * Handlers are allowed to throw, but avoid that because anything that's thrown gets turned
     * into an error that's send to the Flutter side of the bridge.
     *
     * Usage:
     * ```kotlin
     * bridgePort.registerHandlerSuspend("name") { params ->
     *   ...
     *   return someValue
     *   // Or throw when there's an error.
     * }
     * ```
     */
    fun registerHandlerSuspend(
        callbackName: String,
        handler: SuspendBridgeHandlerFunction
    )

    /**
     * Unregister a handler by name.
     */
    fun unregisterHandler(callbackName: String)

    /**
     * Unregister all handlers for this port.
     */
    fun unregisterAllHandlers()
}


/**
 * Custom exception that may be thrown when calling methods on the Flutter side.
 */
class MethodChannelBridgeException private constructor(val reason: MethodChannelBridgeExceptionCause) :
    Exception() {

    companion object {
        @JvmStatic
        fun receiverNotFound(): MethodChannelBridgeException {
            return MethodChannelBridgeException(MethodChannelBridgeExceptionCause.RECEIVER_NOT_FOUND)
        }

        @JvmStatic
        fun bridgeNotInitialized(): MethodChannelBridgeException {
            return MethodChannelBridgeException(MethodChannelBridgeExceptionCause.BRIDGE_NOT_INITIALIZED)
        }

        @JvmStatic
        fun methodExecutionFailed(): MethodChannelBridgeException {
            return MethodChannelBridgeException(MethodChannelBridgeExceptionCause.METHOD_EXECUTION_FAILED)
        }
    }
}

enum class MethodChannelBridgeExceptionCause {
    /**
     * Means that the called method was not found on the Flutter side. This may be because
     * the bridge has not been initialized on the Flutter side.
     */
    RECEIVER_NOT_FOUND,

    /**
     * Means that the method channel bridge has not been initialized yet, and thus cannot be
     * used. You must call the `initializeBridge` method before calling methods on the Flutter
     * side. Although you can register callbacks before the bridge is initialized, code in the
     * Flutter side won't be able to reach those callbacks until initialization.
     */
    BRIDGE_NOT_INITIALIZED,

    /**
     * Means that the call to the Flutter side was made but the Flutter code threw an error.
     */
    METHOD_EXECUTION_FAILED;
}
